### ISSUE TITLE
Orientation enum does not contain upside down

### ISSUE THREAD

**Issue by Trulsmatias** [flutter/flutter#74597]
From https://api.flutter.dev/flutter/widgets/Orientation-class.html.

This is useful because I want to implement functionality similar to Pokemon Go's "Battery Saver" setting, where the screen is dark and non responsive when the device is upside down.

**Comment by pedromassangocode** [comment]
Hi @Trulsmatias 
I'm not sure what you are trying to say. Can you please elaborate what you are trying to do?

If you want to force a custom device rotation you can do so by using SystemChrome, here is a example:

```
void main() {
  WidgetsFlutterBinding.ensureInitialized();
  SystemChrome.setPreferredOrientations([
    DeviceOrientation.portraitDown,
  ]);
  runApp(FlutterApp());
}
```

This will change the orientation of the entire app.

**Comment by Trulsmatias** [comment]
Sorry for the lack of clarity. I do not want to force orientation. I want to detect when the user is holding the device upside down. From the docs this is done (as far as I can see) by OrientationBuilder or MediaQuery.of(context).orientation: https://flutter.dev/docs/cookbook/design/orientation, which both uses the Orientation enum. As far as I can see the DeviceOrientation enum is only used to force and not detect orientation.

**Comment by nt4f04uNd** [comment]
there was similar issue #17215
it was weirdly closed
i don't understand why as this definitely should be a feature of the framework

**Comment by pedromassangocode** [comment]
Hi @Trulsmatias 
As suggested in https://github.com/flutter/flutter/issues/17215 the Flutter team suggest this feature to be a package rather than build it in the core of the Framework.

Looks like the author of https://github.com/flutter/flutter/issues/17215 already made a plugin to get/listen the current device orientation. You can find it in: https://pub.dartlang.org/packages/native_device_orientation.

**Comment by Trulsmatias** [comment]
Unfortunately the plugin is not well enough supported and maintained to be used in production: https://github.com/rmtmckenzie/flutter_native_device_orientation/issues/10. The workaround provided in the issue I linked to does not provide quite the right feel either unfortunately.

This is basic functionality that should be part of the core framework. Please reopen the issue.

**Comment by pedromassangocode** [comment]
> This is basic functionality that should be part of the core framework. Please reopen the issue.

See https://github.com/flutter/flutter/issues/17215#issuecomment-570435794 for reasons. 
Thank you

**Comment by Trulsmatias** [comment]
Thank you for reopening! 🚀 I would really appreciate this feature 😊

**Comment by steeling** [comment]
See #17215 (comment) for reasons.


> > This is basic functionality that should be part of the core framework. Please reopen the issue.
> 
> See [#17215 (comment)](https://github.com/flutter/flutter/issues/17215#issuecomment-570435794) for reasons.
> Thank you

Can you provide a better reason? There's clearly at least some desire from users for this feature to exist, and saying "we want to keep it small" feels a little dismissive.

How different is this than interacting with the device orientation via systemchrome. setPreferredOrientations?

**Comment by ueman** [comment]
Copied from #95719:



## Use case



Flutter supports setting the devices allowed device orientation by using [`SystemChrome.setPreferredOrientations()`](https://api.flutter.dev/flutter/services/SystemChrome/setPreferredOrientations.html). I would like to read the current orientation in order to do something like:

1. Read & save current configuration
2. Temporarily set a different preferred device orientation
3. Reset to the configuration from step 1

## Proposal



Add an API like 
```dart
List<DeviceOrientation> currentPreferredOrientations =  await SystemChrome.getPreferredOrientations();
```

**Comment by stuartmorgan** [comment]
> There's clearly at least some desire from users for this feature to exist

That's true of every package that exists, by definition. If that were the bar, all functionality would be in the framework/engine.

> and saying "we want to keep it small" feels a little dismissive.

I'm not sure why you believe that leveraging the plugin system that's build into Flutter for just this kind of thing is dismissive.

Generally things that control the UI (like preferred orientation) are in the framework, whereas things that are about lower-level device state are left to plugins. It's not clear why this should be different.

> Unfortunately the plugin is not well enough supported and maintained to be used in production

That's not an intrinsic problem with plugins, just that particular implementation. There didn't seem to have been so much demand for this feature that it can't reasonably be left to a plugin.

**Comment by juliansteenbakker** [comment]
I think it would be great if this feature is added in the core framework. In order to rotate the camera the correct way this feature is necessary.

I see it more of an update of the current implementation rather than a new feature since we already can detect if the orientation is landscape of portrait using `MediaQuery.of(context).orientation == Orientation.portrait`.

My suggestion would be to have something like `MediaQuery.of(context).deviceOrientation == DeviceOrientation. portraitDown`, so we won't break current implementations. Another option could be to have an extension on `MediaQuery.of(context).orientation` so we can do something like `MediaQuery.of(context).orientation.details` or something similar that would return the DeviceOrientation.

This way we can not only set the preferred `DeviceOrientation` but also request it.

**Comment by jwang0512** [comment]
facing the same problem, I need the detailed orientation to add padding on camera side only. 

`SafeArea(top: false, bottom: false, right:false, child: Container(color:Colors.pink))`
this works only when camera on left.
![Simulator Screenshot - iPhone 15 Pro - 2024-04-15 at 10 14 17](https://github.com/flutter/flutter/assets/5328540/52e9e808-bd61-40b0-884c-3f7c571bfe63)

but if the camera is on right:
![Simulator Screenshot - iPhone 15 Pro - 2024-04-15 at 10 14 28](https://github.com/flutter/flutter/assets/5328540/8337ded1-b2cb-4bc9-9569-09a52aff92c8)

get a detailed orientation is really needed in this case.

**Comment by Gustl22** [comment]
Isn't the needed functionality already used in `camerax`, see:
https://github.com/flutter/packages/blob/848d7e915e2110984bdb6140e044c232ece00c3a/packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/DeviceOrientationManager.java#L127

So if camerax is using it, the functionality (see https://github.com/flutter/packages/pull/7044), so it must be kind of maintained anyway (?)

Ref: https://developer.android.com/media/camera/camera2/camera-preview#orientation_calculation

I think the mechanism to get the current device rotation must also be used in the native Youtube app, which forces a  device screen rotation on pressing fullscreen in portrait mode, but detects (by leaving the fullscreen mode) if turning the device to landscape and then back to portrait.
